server/polar/order/tasks.py

```diff
@@ -223,3 +232,147 @@ async def process_dunning_order(order_id: uuid.UUID) -> None:
 
         await order_service.process_dunning_order(session, order)
+
+
+@actor(
+    actor_name="order.backfill_order_events",
+    priority=TaskPriority.LOW,
+    time_limit=3600_000,
+)
+async def backfill_order_events(rate_limit_delay: float = 1.0) -> None:
+    """
+    Backfill order.paid and order.refunded events for all existing orders.
+    Uses transaction timestamps for accurate event timing.
+
+    Args:
+        rate_limit_delay: Seconds to wait between batch inserts (default: 1.0)
+    """
+    async with AsyncSessionMaker() as session:
+        existing_order_ids_subquery = (
+            select(Event.user_metadata["order_id"].as_string().label("order_id"))
+            .where(Event.name.in_([SystemEvent.order_paid, SystemEvent.order_refunded]))
+            .distinct()
+            .subquery()
+        )
+
+        statement = (
+            select(Order)
+            .outerjoin(
+                existing_order_ids_subquery,
+                existing_order_ids_subquery.c.order_id == Order.id.cast(String),
+            )
+            .where(
+                Order.deleted_at.is_(None),
+                existing_order_ids_subquery.c.order_id.is_(None),
+            )
+            .options(selectinload(Order.customer))
+            .order_by(Order.created_at.asc())
+        )
+
+        orders_stream = await session.stream_scalars(
+            statement,
+            execution_options={"yield_per": settings.DATABASE_STREAM_YIELD_PER},
+        )
+
+        total_orders = 0
+        total_events = 0
+        events_batch: list[dict[str, Any]] = []
+
+        async for order in orders_stream:
+            events = await _build_events_for_orders(session, [order])
+            events_batch.extend(events)
+            total_orders += 1
+
+            if len(events_batch) >= settings.DATABASE_STREAM_YIELD_PER:
+                await EventRepository.from_session(session).insert_batch(events_batch)
+                total_events += len(events_batch)
+
+                log.info(
+                    "backfill_order_events.progress",
+                    orders=total_orders,
+                    events=total_events,
+                )
+
+                events_batch = []
+                await asyncio.sleep(rate_limit_delay)
+
+        if events_batch:
+            await EventRepository.from_session(session).insert_batch(events_batch)
+            total_events += len(events_batch)
+
+        log.info(
+            "backfill_order_events.completed",
+            total_orders=total_orders,
+            total_events=total_events,
+        )
+
+
+async def _build_events_for_orders(
+    session: AsyncSession, orders: list[Order]
+) -> list[dict[str, Any]]:
+    """Build events for a batch of orders using transaction timestamps."""
+    order_ids = [order.id for order in orders]
+    orders_by_id = {order.id: order for order in orders}
+
+    transactions_result = await session.execute(
+        select(Transaction)
```

Comment:
A bit surprising to see Transaction involved here (Order.created_at / Refund.created_at could have worked). But that's fine :)

Reply:
Haha, I was scratching my head on this. But maybe a classic case of going across the river for the water